library/src/main/scala/giter8/launcher/SbtLaunchJar.scala

```diff
@@ -0,0 +1,165 @@
+package giter8.launcher
+
+import java.io.File
+import java.math.BigInteger
+import java.security.MessageDigest
+
+import gigahorse.support.okhttp.Gigahorse
+import org.apache.commons.io.FileUtils
+
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.util.{Failure, Success, Try}
+
+object SbtLaunchJar {
+
+  import giter8.G8.RichFile
+  import giter8.Tap
+
+  /**
+    * Calculate the SHA1 checksum of an array of bytes
+    *
+    * @param bytes
+    * @return
+    */
+  private def sha1sum(bytes: Array[Byte]): String = {
+    val digest = MessageDigest.getInstance("SHA1").digest(bytes)
+    new BigInteger(1, digest).toString(16).reverse.padTo(40, "0").reverse.mkString
+  }
+
+  private lazy val tmpdir = new File(System.getProperty("java.io.tmpdir")) / "giter8"
+
+  /**
+    * Retrieve the SHA1 checksum for a particular file in Maven Central
+    *
+    * @param url
+    * @return
+    */
+  private def getRemoteSum(url: String): Try[String] = {
+    Gigahorse.withHttp(Gigahorse.config) {
+      http =>
+
+        val request = Gigahorse.url(url + ".sha1").get
+        val response = http.run(request).map(_.bodyAsString)
+
+        // TODO?: Configurable timeout
+        Try(Await.result(response, 1.minute))
+    }
+  }
+
+  /**
+    * Calculate the SHA1 checksum of a file
+    *
+    * @param file
+    * @return
+    */
+  private def getLocalSum(file: File): Try[String] =
+    Try(FileUtils.readFileToByteArray(file)).map(sha1sum)
+
+  /**
+    * Fetch a jar from a url, checking it against the checksum associated
+    * with that url
+    *
+    * @param url
+    * @param file
+    * @param checksum
+    * @return
+    */
+  private def fetchJar(url: String, file: File, checksum: Option[String] = None): Try[File] = {
+    Gigahorse.withHttp(Gigahorse.config) {
+      http =>
+
+        checksum.map(Success.apply).getOrElse(getRemoteSum(url)).flatMap {
+          remoteSum =>
+
+            FileUtils.touch(file)
+            println("Fetching remote jar...")
+            val request = Gigahorse.url(url).get
+            val response = http.run(request).map {
+              r =>
+                val bytes = r.bodyAsByteBuffer.array
+                val localSum = sha1sum(bytes)
+                if (localSum == remoteSum) {
+                  println("Downloaded jar is valid!")
+                  Try(FileUtils.writeByteArrayToFile(file, bytes))
+                    .map(_ => file)
+                } else {
+                  Failure(new RuntimeException("Downloaded jar is invalid... exiting"))
+                }
+            }
+
+            Try(Await.result(response, 10.minutes))
+        }.flatten.tap {
```

Comment:
Is this Rubyism?  If the following Scala is equivalent, I might prefer it because I'd forget what `tap` is.

```scala
}.flatten match {
  case x@Failure(e) =>
    e.printStackTrace()
    FileUtils.forceDelete(file)
    x
  case x => x
}
```